cmd/collector/main.go

```diff
@@ -85,10 +108,27 @@ func main() {
 	r := mux.NewRouter()
 	apiHandler := app.NewAPIHandler(jaegerBatchesHandler, zipkinSpansHandler)
 	apiHandler.RegisterRoutes(r)
+
 	httpPortStr := ":" + strconv.Itoa(*builder.CollectorHTTPPort)
 	recoveryHandler := recoveryhandler.NewRecoveryHandler(logger, true)
-	logger.Info("Listening for HTTP traffic", zap.Int("http-port", *builder.CollectorHTTPPort))
-	if err := http.ListenAndServe(httpPortStr, recoveryHandler(r)); err != nil {
-		logger.Fatal("Could not launch service", zap.Error(err))
+	logger.Info("Starting Jaeger Collector HTTP server", zap.Int("http-port", *builder.CollectorHTTPPort))
+
+	go func() {
+		// there's still a chance that a request comes in between the `ready` flag and the server's
+		// actual startup, but this would be a rare situation, as we assume the `ListenAndServe` is fast enough
+		// between allocating a port and serving a request.
+		serverState = http.StatusNoContent
+
+		logger.Fatal("Could not launch service", zap.Error(http.ListenAndServe(httpPortStr, recoveryHandler(r))))
+
+		// if we returned from the call above, it's because we have troubles... the logger.Fatal should panic,
+		// so, this code probably won't be reached, but if we are hung for some reason, we can still notify our
+		// state to an external actor
+		serverState = http.StatusInternalServerError
+	}()
+
+	select {
+	case <-serverChannel:
+		logger.Info("Jaeger Collector is finishing")
```

Comment:
I checked both OpenShift and Kubernetes and they work on top of Cobra. So, a Cobra command for "all in one" (OpenShift here), would call the "StartMaster" command, which would then start that component.

In our case, if we were to take OpenShift as example, it would look like this:

- Single binary for the whole application (not required, but I see some advantages to it)
- Commands passed to the binary specify what to do (`jaeger start-all-in-one` or even `jaeger start all-in-one`)
- The Cobra's "`Run`" for the top-level commands would invoke specialized commands, like, `jaeger start all-in-one` means `pkg/cmd/hcOptions.RunHealthCheckServer` + `pkg/cmd/collectorOptions.RunCollector` + `pkg/cmd/agentOptions.RunAgent`, ... (reference: https://github.com/openshift/origin/blob/master/pkg/cmd/server/start/start_allinone.go#L286) .
- With this, the main file itself is quite clean: https://github.com/openshift/origin/blob/master/cmd/openshift/openshift.go